triplet/main.py

```diff
@@ -0,0 +1,177 @@
+"""
+    PyTorch training code for TFeat shallow convolutional patch descriptor:
+    http://www.iis.ee.ic.ac.uk/%7Evbalnt/shallow_descr/TFeat_paper.pdf
+
+    The code reproduces *exactly* it's lua anf TF version:
+    https://github.com/vbalnt/tfeat
+
+    2017 Edgar Riba
+"""
+
+from __future__ import print_function
+import argparse
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+from torch.autograd import Variable
+
+from PIL import Image
+import numpy as np
+
+# Training settings
+parser = argparse.ArgumentParser(description='PyTorch TFeat Example')
+parser.add_argument('--imageSize', type=int, default=32,
+                    help='the height / width of the input image to network')
+parser.add_argument('--batch-size', type=int, default=64, metavar='N',
+                    help='input batch size for training (default: 64)')
+parser.add_argument('--n_triplets', type=int, default=6400, metavar='N',
+                    help='how many triplets will generate from the dataset')
+parser.add_argument('--epochs', type=int, default=10, metavar='N',
+                    help='number of epochs to train (default: 2)')
+parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
+                    help='learning rate (default: 0.01)')
+parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
+                    help='SGD momentum (default: 0.5)')
+parser.add_argument('--no-cuda', action='store_true', default=False,
+                    help='enables CUDA training')
+parser.add_argument('--seed', type=int, default=1, metavar='S',
+                    help='random seed (default: 1)')
+parser.add_argument('--log-interval', type=int, default=10, metavar='N',
+                    help='how many batches to wait before logging training status')
+
+args = parser.parse_args()
+args.cuda = not args.no_cuda and torch.cuda.is_available()
+
+torch.manual_seed(args.seed)
+if args.cuda:
+    torch.cuda.manual_seed(args.seed)
+
+
+class TripletMNIST(datasets.MNIST):
+    """From the MNIST Dataset it generates triplet samples
+    note: a triplet is composed by a pair of matching images and one of
+    different class.
+    """
+    def __init__(self, *arg, **kw):
+        super(TripletMNIST, self).__init__(*arg, **kw)
+
+        print('Generating triplets ...')
+        self.n_triplets = args.n_triplets
+        self.train_triplets = self.generate_triplets(self.train_labels)
+
+    def generate_triplets(self, labels):
+        triplets = []
+        for x in xrange(self.n_triplets):
+            idx = np.random.randint(0, labels.size(0))
+            idx_matches = np.where(labels.numpy() == labels[idx])[0]
+            idx_no_matches = np.where(labels.numpy() != labels[idx])[0]
```

Comment:
You could compute these two arrays for every class outside of the loop. That would likely make it faster.